ami/flowchart/library/Scipy.py

```diff
@@ -194,6 +224,105 @@ def addInput(self, **args):
         def to_operation(self, **kwargs):
             return gn.Map(name=self.name()+"_operation", **kwargs, func=FitProc(**self.values))
 
+
+
+    class FitPeakProc():
+
+        def __init__(self, *args, **kwargs):
+            self.model = kwargs['Model']
+            self.use_offset = kwargs['Use offset']
+            self.a_0 = kwargs['Initial amplitude']
+            self.x_0 = kwargs['Initial x0']
+            self.fwhm_0 = kwargs['Initial FWHM']
+            self.c_0 = kwargs['Initial offset']
+            self.func = None
+
+        def get_func(self):
+            if self.model == "Gaussian":
+                sigma_0 = self.fwhm_0 / 2.355
+                if self.use_offset:
+                    p0 = [self.a_0, self.x_0, sigma_0, self.c_0]
+                    func = lambda x, a, mu, sig, c: gaussian_func(x, a, mu, sig) + c
+                    return func, p0
+                else:
+                    p0 = [self.a_0, self.x_0, sigma_0]
+                    return gaussian_func, p0
+
+            elif self.model == "Lorentzian":
+                gamma_0 = self.fwhm_0 / 2
+                if self.use_offset:
+                    p0 = [self.a_0, self.x_0, gamma_0, self.c_0]
+                    func = lambda x, a, x0, gamma, c: lorentzian_func(x, a, x0, gamma) + c
+                    return func, p0
+                else:
+                    p0 = [self.a_0, self.x_0, gamma_0]
+                    return lorentzian_func, p0
+
+            elif self.model == "Moments":
+                return stats_from_moments, None
+            return
+
+        def __call__(self, y, *args, **kwargs):
+            if self.func is None:
+                self.func, self.p0 = self.get_func()
+
+            x = np.arange(0, y.size, 1)
+
+            if self.p0 is None:
+                # Calculate moments and make-up a gaussian
+                # Dirtier but faster
+                mean, sigma, skew = self.func(x, y=y)
+                fwhm = 2.355 * sigma
+                ampl = np.max(y)
+                y = gaussian_func(x, ampl, mean, sigma)
+                return y, ampl, mean, sigma, fwhm, 0.0
+
+            try:
+                # Real fits
+                best_vals, covar = optimize.curve_fit(self.func, x, y, p0=self.p0)
+                if self.model == "Gaussian":
+                    fwhm = 2.355 * best_vals[2]
+                elif self.model == "Lorentzian":
+                    fwhm = 2 * best_vals[2]
+
+                if self.use_offset:
+                    return self.func(x, *best_vals), best_vals[0], best_vals[1], best_vals[2], fwhm, best_vals[3]
+                else:
+                    return self.func(x, *best_vals), best_vals[0], best_vals[1], best_vals[2], fwhm, 0.0
+            except RuntimeError:
+                printExc()
```

Comment:
I was thinking about this when I was working on the CurveFit box today. If a fit fails, should we show a warning in the client? I'm not sure if this currently does?

Reply:
Not quite indeed. I was wondering if doing a generic try/except would be ok?
The type of errors I've seen while testing should probably mostly be ignored / turned into a warning. For example, if there aren't enough point yet (in a scan for example), it will fail because there are more parameters than data points.

Reply:
This should do what we want I think: https://github.com/slac-lcls/ami/pull/91